Fix circular deps

tcojean opened the pull request:
This PR fixes detected circular dependencies and improves the CI.

The approach used here is to keep the current system and enforce that no kernel module ever use any function with implementation in the core module. That was the case for two things in particular, the static variables of the `CudaExecutor` and the GMRES `solve_x` function. The `jacobi`  CUDA `advanced_apply` is borderline, I am somehow not able to make it throw an error. In fact, the CI is able to find some problem with the Jacobi (CUDA 9.2, GCC 7, Release build only), so it also requires fixing.

I really need the confirmation of @hartwiganzt here (or any mac user): does this build with `-DBUILD_SHARED_LIBS=ON`?

### Changes
+ Add a CI step to detect circular dependencies between Ginkgo modules.
+ GMRES used dense matrices and uses the preconditioner apply. These calls are
  moved to the core module
+ The static variables for the CudaExecutor were placed in the core module. They
  are moved to the Cuda module.
+ Fix the jacobi circular dependency using some code duplication.
   + All the `apply` kernels and the related kernel selector are duplicated with an
      alpha parameter in the calls
   + The advanced version scale by alpha inside the kernel themselves.
   + Use a new `multiply_vec_add` instead of `multiply_vec` which preserves the
      value in `x` by storing with addition.
+ Reduce the amount of CI jobs and integrate static builds.
  + For this, we vary more than one parameter at once. Here is the new overview:
  + core + gcc + debug + static
  + core + clang + release + shared
  + omp + gcc + release + shared
  + omp + clang + debug + static
  + all + gcc + cuda 9.0 + release + shared
  + all + clang + cuda 9.0 + debug + static
  + all + gcc + cuda 9.1 + release + static
  + all + clang + cuda 9.1 + debug + shared
  + all + gcc + cuda 9.2 + debug + shared
  + all + clang + cuda 9.2 + release + static
  + all + gcc + cuda 10.0 + debug + static
  + all + clang + cuda 10.0 + release + shared

### Closes #203.
### Closes #270.

### TODO
+ [x] Add a CI step which should fail when Ginkgo has circular dependencies
+ [x] Fix the existing circular dependencies
+ [x] Add a static library build to the CI system and try to trim the amount of configurations by varying more than one parameter per configuration
+ [x] Add a note to the contributing guideline on this circular dependency problem
+ [x] Fix Jacobi circular dependency

tcojean commented:
This how it looks like when there are circular dependencies in Ginkgo:
https://gitlab.com/ginkgo-project/ginkgo-public-ci/pipelines/52941979

The pipeline step `no-circular-deps` fails with the detected problems, which are then fixed in the next commit.

thoasm commented on cuda/solver/gmres_kernels.cu:
```diff
@@ -720,15 +714,6 @@ void solve_x(std::shared_ptr<const CudaExecutor> exec,
-        initialize<matrix::Dense<ValueType>>({one<ValueType>()}, exec);
-    x->add_scaled(one_op.get(), after_preconditioner.get());
-    // Solve x
-    // x = x + preconditioner_ * krylov_bases * y
```
I think the `Solve x` comment can stay (we might have to replace it with `Solve for before_preconditioner` or similar).

tcojean replied:
Welle there is no more `preconditioner` here, so it's not really the same operation anymore, we only have krylov_bases * y I believe?

Also, I will change the name of this function to `calculate_qy` to beging with, as it is no more `solve_x`.

thoasm commented on cuda/solver/gmres_kernels.cu:
```diff
@@ -688,17 +688,11 @@ void solve_upper_triangular(
              const matrix::Dense<ValueType> *y, matrix::Dense<ValueType> *x,
-             const Array<size_type> *final_iter_nums,
-             const LinOp *preconditioner)
+             matrix::Dense<ValueType> *before_preconditioner,
```
I think you can remove the `x` parameter from here as you don't need it inside the function anymore.

thoasm commented on cuda/solver/gmres_kernels.cu:
```diff
@@ -738,12 +723,12 @@ void step_2(std::shared_ptr<const CudaExecutor> exec,
             const matrix::Dense<ValueType> *krylov_bases,
             const matrix::Dense<ValueType> *hessenberg,
             matrix::Dense<ValueType> *y, matrix::Dense<ValueType> *x,
```
I think you don't need the `x` here either, so you can probably remove it from the signature of `step_2`.

thoasm commented:
@tcojean I am not sure if it related to this, but I remember resolving similar problems when I fixed a linker error with test cases in #199, where I brute forced the linker to accept linking against all ginkgo libraries again. It also might have been an issue with circular dependencies, so maybe we can remove these additional link targets with this PR.

hartwiganzt commented:
Works fine on MacOS

tcojean commented:
@hartwiganzt just to confirm, with shared libraries, i.e. `-DBUILD_SHARED_LIBS=ON` ?

hartwiganzt commented:
Confirm.

tcojean commented:
I also added to the MR a reduction of the amount of CI jobs (six less) and added CI jobs for static builds. I do this by shuffling around (manually) the parameters between the different builds so that we try to test as much as possible overall without having one job for each single parameter change. This is also now more balanced compared to before where we had a lot of debug tests compared to the rest.

yhmtsai commented:
it also works fine on my MacOS, with default settings but `no benchmark`, `no examples`.
But it seems to be failed on the CI.

tcojean commented:
@yhmtsai thanks for the input Mike.

Indeed, the CI fails because there is still one circular dependency left in jacobi. Weirdly enough, it is only triggered in some particular case (clang 9.2, gcc 7.x, release build, ...), but it is there. I will fix this and this should be able to be merged afterwards.

tcojean commented:
The CI will hopefully confirm, but the block jacobi circular dependency should now be fixed.

Here is a summary of what I did:

Fix the jacobi circular dependency using some code duplication.

+ All the `apply` kernels and the related kernel selector is duplicated with an
	alpha call
+ The advanced version scale by alpha inside the kernel
+ Use a new `multiply_vec_add` instead of `multiply_vec` which preserves the
	value in `x` by storing with addition.

thoasm commented on cuda/preconditioner/jacobi_kernels.cu:
```diff
@@ -277,6 +278,39 @@ __global__ void __launch_bounds__(warps_per_block *cuda_config::warp_size)
 
+template <int max_block_size, int subwarp_size, int warps_per_block,
+          typename ValueType, typename IndexType>
+__global__ void __launch_bounds__(warps_per_block *cuda_config::warp_size)
```
Very small nit: It looks like the `cuda_config::warp_size` is a pointer instead of a multiplication with `warps_per_block`.
```suggestion
__global__ void __launch_bounds__(warps_per_block * cuda_config::warp_size)
```

tcojean replied:
cmake-format is the one who does that for some reason

tcojean replied:
And I think it does think that this is a pointer, my linter also shows it exactly in that way.

thoasm replied:
Then I guess we leave it as it is, it is definitively not a big deal, just looked a bit off to me.
I definitively don't think it is worth looking into the style settings for clang-format to change such a small detail, so I will mark this as 'Resolved'.

thoasm commented on cuda/components/warp_blas.cuh:
```diff
@@ -354,6 +354,42 @@ __device__ __forceinline__ void multiply_vec(
+    int max_problem_size, typename Group, typename MatrixValueType,
+    typename VectorValueType,
+    typename = xstd::enable_if_t<group::is_communicator_group<Group>::value>>
+__device__ __forceinline__ void multiply_vec_add(
```
If you want to remove the code duplication, you could do something similar to [the reduce function in `ginkgo/cuda/components/reduction.cuh`](https://github.com/ginkgo-project/ginkgo/blob/develop/cuda/components/reduction.cuh#L66), which is using a templated lambda function to define the last step (maybe calling it `Closure closure_op`), and do at the very end:
```
closure_op(res[group.thread_rank() * res_increment] , out);
```

You could make a function that does exactly that, and then call it in both `multiply_vec` and `multiply_vec_add` with a different lambda function. For `multiply_vec_add`, the call would look something like that:
```
multiply_vec_lambda(group, problem_size, vec, mtx_row, mtx_increment, res, res_increment,
                    [](VectorValueType &result, const VectorValueType &out) {result += out;});
```

tcojean replied:
Good idea, I'll do that.

tcojean replied:
Everything should now be fixed, including this. Thanks for the comments.

thoasm commented on cuda/preconditioner/jacobi_kernels.cu:
```diff
@@ -634,14 +743,21 @@ void apply(std::shared_ptr<const CudaExecutor> exec, size_type num_blocks,
+            syn::value_list<int, cuda_config::min_warps_per_block>(),
+            syn::type_list<>(), num_blocks, block_precisions.get_const_data(),
+            block_pointers.get_const_data(), blocks.get_const_data(),
+            storage_scheme, alpha->get_const_values(),
```
Since you are using `alpha[0]` inside the `apply` kernel, you also need to shift the `alpha` pointer in order to get the correct one as far as I know.
```suggestion
            storage_scheme, alpha->get_const_values() + col,
```

tcojean replied:
Currently, I think alpha expects to always be a scalar, ever for the multiple RHS case.

See in `cuda/test/preconditioner/jacobi_kernels.cpp` the test `CudaLinearCombinationApplyToMultipleVectorsEquivalentToRef`. Cuda is a lone scalar.

thoasm replied:
I think both cases can be the case, I only know that in the solver (like here in [`core/solver/bicgsab.cpp`](https://github.com/ginkgo-project/ginkgo/blob/develop/core/solver/bicgstab.cpp#L86)) we have as many `alpha`s as we have right hand sides.
I will look if we documented that anywhere.

thoasm replied:
It seems that you are right, I might have confused it (the `alpha` in the solver is probably not the same as the one in `apply`).
In [the advanced `apply` in LinOp](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/lin_op.hpp#L527), we use `validate_application_parameters` (defined [here](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/lin_op.hpp#L283)), which checks if `alpha` and `beta` have `dim<2>(1,1)`.

So you are correct that it is always a `1x1` matrix.

tcojean replied:
I did see implementations though which work with `alpha[col]`or `alpha[alpha_id]`. That is the case for CGS and some Dense kernels. There may actually be a problem there ?

thoasm replied:
In the solvers, I was wrong, they have a separate `apply_impl` (like [here in Bicgstab](https://github.com/ginkgo-project/ginkgo/blob/develop/core/solver/bicgstab.cpp#L176)) to handle those `alpha` and `beta`. I just thought `alpha` always meant the same, which is clearly not the case (and was kind of naive from me).

tcojean replied:
OK and in dense it's purely for scaling where alpha has the meaning of what you scale by, so that is normal also.

thoasm commented on cuda/components/warp_blas.cuh:
```diff
@@ -327,13 +327,13 @@ __device__ __forceinline__ void multiply_transposed_vec(
     const VectorValueType &__restrict__ vec,
     const MatrixValueType *__restrict__ mtx_row, uint32 mtx_increment,
-    VectorValueType *__restrict__ res, uint32 res_increment)
+    VectorValueType *__restrict__ res, uint32 res_increment, Closure closure_op)
```
Could you also document this Closure in the doxygen above? Something like
```
@tparam Closure  type of the function used as to write the result
[...]
@param closure_op  Operation that is performed when writing to
                   `res[group.thread_rank() * res_increment]` as 
                   `closure_op(res[group.thread_rank() * res_increment], out)`
                   where `out` is the result of \f$ res = mtx \cdot vec \f$.
```
or similar.

tcojean replied:
Oops, thanks for pointing that out. This is fixed.
